scm/etc/scripts/UFS_IC_generator.py

```diff
@@ -694,1391 +1304,1436 @@ def get_UFS_surface_data(dir, tile, i, j, old_chgres):
+        # Specific humidity @ time > 0
+        qv_rev[0,:] = qv_lay[t,::-1]
+        for k in range(0,nlevs): dp2[0,k] = to_p[0,k+1] - to_p[0,k]
+        qv_rev_new = fv3_remap.map1_q2(nlevs, from_p, qv_rev, nlevs, to_p, dp2,        \
+                                       0, 0, 0, kord_tr, q_min)
+        # Zonal wind  @ time > 0
+        u_rev[0,:] = u_lay[t,::-1]
+        u_rev_new  = fv3_remap.map1_ppm(nlevs, from_p, u_rev, 0.0, nlevs, to_p,        \
+                                        0, 0, -1, kord_tm )
+        # Meridional wind @ time > 0
+        v_rev[0,:] = v_lay[t,::-1]
+        v_rev_new  = fv3_remap.map1_ppm(nlevs, from_p, v_rev, 0.0, nlevs, to_p,        \
+                                        0, 0, -1, kord_tm )
+
+        # Store
+        p_layr[t+1,:]  = p_lay[t+1,:]
+        p_levr[t+1,:]  = p_lev[t+1,:]
+        tv_layr[t+1,:] = tv_rev_new[0,::-1]
+        qv_layr[t+1,:] = qv_rev_new[0,::-1]
+        u_layr[t+1,:]  = u_rev_new[0,::-1]
+        v_layr[t+1,:]  = v_rev_new[0,::-1]
+
+    #
+    p_layr[t+2,:]  = p_layr[t+1,:]
+    p_levr[t+2,:]  = p_levr[t+1,:]
+    tv_layr[t+2,:] = tv_layr[t+1,:]
+    qv_layr[t+2,:] = qv_layr[t+1,:]
+    u_layr[t+2,:]  = u_layr[t+1,:]
+    v_layr[t+2,:]  = v_layr[t+1,:]
+
+    # Temperature
+    t_layr = tv_layr/(1.0 + zvir*qv_layr)
+
+    # Create dictionary with "Regridded" state
+    stateREGRID = {"time":   np.concatenate((np.asarray([0.]), time_hr[:])),           \
+                   "ps":     np.concatenate((ps[0:1], ps), axis=0),                    \
+                   "p_lev":  p_levr,  \
+                   "p_lay":  p_layr,  \
+                   "t_lay":  t_layr,  \
+                   "qv_lay": qv_layr, \
+                   "u_lay":  u_layr,  \
+                   "v_lay":  v_layr,  \
+                   "tv_lay": tv_layr  }
+    
+    ####################################################################################
+    #
+    # Compute tendencies advective = total - remapping
+    # 
+    ####################################################################################
+
+    #
+    dtdt_adv   = np.zeros([n_files+1,nlevs])
+    dqvdt_adv  = np.zeros([n_files+1,nlevs])
+    dudt_adv   = np.zeros([n_files+1,nlevs])
+    dvdt_adv   = np.zeros([n_files+1,nlevs])
+    pres_adv   = np.zeros([n_files+1,nlevs])
+    pres_i_adv = np.zeros([n_files+1,nlevs+1])
+    tend_remap = np.zeros([1,nlevs])
+    tend_total = np.zeros([1,nlevs])
+
+    #
+    for t in range(n_files):
+        #
+        dtime_sec = (stateNATIVE["time"][t+1] - stateNATIVE["time"][t])*sec_in_hr
+        #
+        pres_adv[t,:]   = stateNATIVE["p_lay"][t,:]
+        pres_i_adv[t,:] = stateNATIVE["p_lev"][t,:]
+        #
+        tend_total[0,:] = stateREGRID["t_lay"][t+1,:] - stateREGRID["t_lay"][t,:]
+        tend_remap[0,:] = stateREGRID["t_lay"][t,:]   - stateNATIVE["t_lay"][t,:]
+        dtdt_adv[t,:]   = (tend_total[0,:] - tend_remap[0,:]) / dtime_sec
```

Comment:
IMO, this method should be revised in the (near) future. Calculating tendencies this way includes the effect of physics already, which we want the physics in the SCM to represent. I think the "right" way to calculate advective tendencies, if we don't want to rely on having either the non-physics tendency terms or the all-physics tendency terms as "special" diagnostics (which most runs will not save due to space constraints), is to actually use the output winds with neighboring profiles of T, q with the upwind differencing scheme to derive them. I'll go ahead and start an issue so that we don't forget to do this.

Reply:
You are correct, I think of this as the quick and easier(dirtier) way.
I'm out of my lane when it comes to numerical recipes for computing advection, but it's not clear to me how using neighboring points gets around the implicit physics influence on the state? (I need to learn more)